json/marshal.go

```diff
@@ -0,0 +1,193 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+const FORCE_SEND_FIELD_NAME = "ForceSendFields"
+
+// Marshal returns a JSON encoding of the given object. Included fields:
+// - non-empty value
+// - a basic type whose field's name is present in forceSendFields
+// Our templates always populate either omitempty or the '-' tag.
+// For simplicity, we assume one of those cases.
+func Marshal(object interface{}) ([]byte, error) {
+
+	dataMap, err := structAsMap(object)
```

Comment:
This is not recursive at the moment. Can we make it recursive?

* Upside: we don't need a custom MarshalJSON function on all our types anymore
* Downside: custom MarshalJSON is not honored/cannot be used

Reply:
The downside is a bit more tricky than it seems. 
* If you don't honor MarshalJSON and do it yourself, you need to suddenly care about many more types. E.g., an array of Jobs needs to be iterated and recursively called for each element. Currently, we call json.Marshall on the array and json figures out it needs to use the custom marshaller on each element.
* All calls to json.Marhsal need to use our library. If  a struct has a dependency on the SDK, it will need to use our marshaler, even if the struct itself has nothing to do with it. 
  * We need to be sure to support many more cases to ensure everything works.
  * Users will have to call our custom marshaller directly too if they build something which requires to marshal elements.